### ISSUE TITLE
Ambiguous operator== between OptionalArrayRef and HeaderOnlyArrayRef breaks plain ArrayRef == ArrayRef on MSVC

### ISSUE THREAD

**Issue by schall8** [pytorch/pytorch#193589]
### 🐛 Describe the bug

Description (PyTorch's issue template uses a ## 🐛 Describe the bug / ### Versions structure — this fits it):

## 🐛 Describe the bug

`c10/util/OptionalArrayRef.h` defines a hidden-friend comparison operator:

```cpp
// c10/util/OptionalArrayRef.h, line 223
friend bool operator==(OptionalArrayRef a1, ArrayRef<T> other) {
  if (!a1.has_value()) {
    return false;
  }
  return a1.value() == other;
}

OptionalArrayRef<T> has an implicit (non-explicit) converting constructor from ArrayRef<T>, so this operator is a viable candidate whenever both operands are plain ArrayRef<T> — the left operand converts to OptionalArrayRef<T> via one user-defined conversion.

Separately, torch/headeronly/util/HeaderOnlyArrayRef.h defines:

// torch/headeronly/util/HeaderOnlyArrayRef.h, line 251
friend bool operator==(HeaderOnlyArrayRef a1, HeaderOnlyArrayRef a2) {
  return a1.equals(a2);
}

ArrayRef<T> converts to HeaderOnlyArrayRef<T>, so this operator is also a viable candidate for two plain ArrayRef<T> operands.                  
With MSVC, these two candidates tie during overload resolution, producing a hard compile error for completely ordinary code such as .sizes() == .sizes(), which appears throughout torch/csrc/api/include/tor

torch/csrc/api/include/torch/nn/functional/embedding.h(103):  "==" matches these operands:
            function "c10::operator==(c10::HeaderOnlyArrayRef<int64_t>, c10::HeaderOnlyArrayRef<int64_t>)" (declared at line 251 of
.../torch/headeronly/util/HeaderOnlyArrayRef.h)
            function "c10::operator==(c10::OptionalArrayRef<int64_t>, c10::ArrayRef<int64_t>)" (declared at line 223 of .../c10/util/OptionalArrayRef.h)
            operand types are: c10::IntArrayRef == c10::IntAr
    if (!(!per_sample_weights_.defined() || input_.sizes() == per_sample_weights_.sizes())) { ...

The same ambiguity also fires in torch/nn/functional/activation.h (key.sizes() == value.sizes()) and torch/nn/functional/loss.h, and in any downstream
code that transitively includes these headers and does a plaiison — this isn't specific to any one project. It reproducedfor us building SageAttention (https://github.com/thu-ml/SageAttention) (csrc/fused/fused.cu, which just does #include <torch/extension.h>) as a CUDA
extension on Windows.

Repro'd with both MSVC 19.44 (VS2022 BuildTools 14.44.35207) 30133), so it's independent of MSVC toolset version. Triggered both compiling a .cu file via nvcc and a plain .cpp compiled directly with cl.exe.

### Versions

(comfy) D:\comfyui\ComfyUI\custom_nodes>curl -sL https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py | python
Collecting environment information...
PyTorch version: 2.13.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro (10.0.26200 64-bit)
GCC version: (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders, r8) 13.2.0
Clang version: Could not collect
CMake version: version 4.4.2
Libc version: N/A

Python version: 3.13.14 | packaged by Anaconda, Inc. | (main, Jul  9 2026, 14:26:41) [MSC v.1942 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26200-SP0
Is CUDA available: True
CUDA runtime version: 13.3.73
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 5070 Ti
Nvidia driver version: 595.97
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Name: AMD Ryzen 9 9900X 12-Core Processor
Manufacturer: AuthenticAMD
Family: 107
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 4400
MaxClockSpeed: 4400
L2CacheSize: 12288
L2CacheSpeed: None
Revision: 17408

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] onnxruntime==1.28.0
[pip3] open_clip_torch==3.3.0
[pip3] optree==0.19.1
[pip3] pytorch-lightning==2.6.5
[pip3] rotary-embedding-torch==0.9.1
[pip3] torch==2.13.0+cu130
[pip3] torch-complex==0.4.4
[pip3] torchaudio==2.11.0+cu130
[pip3] torchcodec==0.16.0
[pip3] torchmetrics==1.9.0
[pip3] torchsde==0.2.6
[pip3] torchvision==0.28.0+cu130
[pip3] triton-windows==3.7.1.post27
[conda] numpy                     2.4.4                    pypi_0    pypi
[conda] open-clip-torch           3.3.0                    pypi_0    pypi
[conda] optree                    0.19.1                   pypi_0    pypi
[conda] pytorch-lightning         2.6.5                    pypi_0    pypi
[conda] rotary-embedding-torch    0.9.1                    pypi_0    pypi
[conda] torch                     2.13.0+cu130             pypi_0    pypi
[conda] torch-complex             0.4.4                    pypi_0    pypi
[conda] torchaudio                2.11.0+cu130             pypi_0    pypi
[conda] torchcodec                0.16.0                   pypi_0    pypi
[conda] torchmetrics              1.9.0                    pypi_0    pypi
[conda] torchsde                  0.2.6                    pypi_0    pypi
[conda] torchvision               0.28.0+cu130             pypi_0    pypi
[conda] triton-windows            3.7.1.post27             pypi_0    pypi


cc @malfet @zklaus @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @nkhasbag-nv

**Comment by schall8** [comment]
# `c10::OptionalArrayRef`'s `operator==(OptionalArrayRef, ArrayRef<T>)` creates ambiguity for plain `ArrayRef == ArrayRef` comparisons

## Bug description

`c10/util/OptionalArrayRef.h` defines a hidden-friend comparison operator:

```cpp
// c10/util/OptionalArrayRef.h, line 223
friend bool operator==(OptionalArrayRef a1, ArrayRef<T> other) {
  if (!a1.has_value()) {
    return false;
  }
  return a1.value() == other;
}
```

`OptionalArrayRef<T>` has an implicit (non-`explicit`) converting constructor from `ArrayRef<T>`, so this operator is a viable candidate whenever *both* operands are plain `ArrayRef<T>` — the left operand converts to `OptionalArrayRef<T>` via one user-defined conversion.

Separately, `torch/headeronly/util/HeaderOnlyArrayRef.h` defines:

```cpp
// torch/headeronly/util/HeaderOnlyArrayRef.h, line 251
friend bool operator==(HeaderOnlyArrayRef a1, HeaderOnlyArrayRef a2) {
  return a1.equals(a2);
}
```

`ArrayRef<T>` derives from / converts to `HeaderOnlyArrayRef<T>`, so this operator is *also* a viable candidate for two plain `ArrayRef<T>` operands.

With MSVC (repro'd on both VS2022 14.44.35207 and VS2019 14.29.30133), these two candidates tie during overload resolution, producing a hard compile error for completely ordinary code such as `some_array_ref == some_other_array_ref` — e.g. `.sizes() == .sizes()`, which appears throughout `torch/csrc/api/include/torch/nn/functional/*`:

```
torch/csrc/api/include/torch/nn/functional/embedding.h(103): error: more than one operator "==" matches these operands:
            function "c10::operator==(c10::HeaderOnlyArrayRef<int64_t>, c10::HeaderOnlyArrayRef<int64_t>)" (declared at line 251 of .../torch/headeronly/util/HeaderOnlyArrayRef.h)
            function "c10::operator==(c10::OptionalArrayRef<int64_t>, c10::ArrayRef<int64_t>)" (declared at line 223 of .../c10/util/OptionalArrayRef.h)
            operand types are: c10::IntArrayRef == c10::IntArrayRef
    if (!(!per_sample_weights_.defined() || input_.sizes() == per_sample_weights_.sizes())) { ...
```

The same ambiguity also fires in `torch/nn/functional/activation.h` (`key.sizes() == value.sizes()`) and `torch/nn/functional/loss.h`, and in any downstream code that transitively includes these headers and does a plain `ArrayRef == ArrayRef` comparison — this is not specific to any one project.

## Environment

- PyTorch: `2.13.0+cu130` (git `cf30153c4c131c8164ee7798e5022d810682e2cb`)
- CUDA: 13.0
- Python: 3.13.14
- OS: Windows 11
- Compiler: MSVC 19.44 (VS2022 BuildTools 14.44.35207) — also reproduced on MSVC 19.29 (VS2019 14.29.30133), so it's independent of MSVC toolset version
- Triggered when compiling a `.cu` file (via nvcc) that transitively includes `<torch/extension.h>`; also affects plain `.cpp` translation units compiled directly with `cl.exe`

## Repro

Any `.cu`/`.cpp` file that does:

```cpp
#include <torch/extension.h>
```

and transitively instantiates `torch::nn::functional::embedding_bag` (or `multi_head_attention_forward`, or several loss functions) will hit this. In practice this reproduced building [SageAttention](https://github.com/thu-ml/SageAttention) (`csrc/fused/fused.cu`, which does `#include <torch/extension.h>`) against this torch build on Windows.

## Suggested fix

Constrain `OptionalArrayRef`'s comparison operator so it isn't a candidate when the left operand is *already* an exact `ArrayRef<T>` — e.g. require the left argument to genuinely be `OptionalArrayRef<T>` rather than accepting it by value (which permits the implicit conversion), or add an `enable_if`/`requires` clause that excludes this overload from participating when a same-rank `HeaderOnlyArrayRef`/`ArrayRef` comparison already applies. Locally, we worked around this by removing the `OptionalArrayRef == ArrayRef` overload entirely, since our use case never compares the two: no other legitimate call site should be affected, since a genuine `OptionalArrayRef` compared against an `ArrayRef` will still need this operator, so the real fix should keep that path working while excluding the plain `ArrayRef == ArrayRef` case.
